test/inductor/test_scatter_optimization.py

```diff
@@ -0,0 +1,118 @@
+# Owner(s): ["module: inductor"]
+
+import copy
+import os
+
+import torch
+from torch import nn
+from torch._dynamo.utils import same
+from torch._inductor import metrics
+from torch._inductor.runtime.runtime_utils import do_bench_gpu as do_bench
+from torch._inductor.test_case import TestCase
+from torch.testing._internal.inductor_utils import HAS_GPU
+
+DO_PERF_TEST = os.environ.get("DO_PERF_TEST") == "1"
+
+
+class TestScatterOpt(TestCase):
+    def setUp(self):
+        super().setUp()
+        metrics.reset()
+
+    def check_metric(self, val=1):
+        self.assertEqual(val, metrics.num_matches_for_scatter_upon_const_tensor)
```

Comment:
Can we just compare metrics against the `num_bytes_accessed`?

Reply:
`num_bytes_accessed` does not work correctly for mutation. We have a StarDep injected to maintain write-after-write dependencies. That StarDep should not contribute any memory access but we count it when estimate number of memory access.

In spite of the issue mentioned, I think we can still use 'num_bytes_accessed' to verify if the optimization kicks in . When the optimization kicks in, we still access less memory even though the sparse matrix is saved to a buffer in the unit test.

Reply:
I've added assertion for `metrics.num_bytes_accessed` for majority of test cases.

But for `test_cross_entropy_loss` , it's non-trivial to estimate the amount of memory access manually. And I don't like putting the printed number in unit test directly either since that does not help with understanding.

I think we can probably keep both:
1. have assertion for num_bytes_accessed
2. have a specific metric tracking if the optimization is triggered. This is especially useful when the model is a bit complex. It's also useful in production models if we want to know if specific optimizations get triggered. (probably need improve tlparse to expose metrics)

WDYT?

Reply:
@shunting314 Don't we already have some infra for checking when patterns get triggered? I think we should be able to reuse that.

Reply:
Do you mean something like this:
```
            self.assertEqual(counters["inductor"]["pattern_matcher_count"], count)
            self.assertEqual(counters["inductor"]["pattern_matcher_nodes"], nodes)
        
```

Reply:
@shunting314 yeah.